### PR TITLE
Move solver to generic class outside of TabOptimize

### LINKED ISSUE #945: Build complete notification is triggered per-thread
https://discord.com/channels/785153694478893126/1071435922597490688

### PR DESCRIPTION
**Pull request by tooflesswulf**
Created a generalized SolverBase class to pull optimization code out of TabOptimize/index as much as possible.

Without #921 this time.

Functionality should be identical to live.

Resolve #945

### CODE REVIEW

File: apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabOptimize/index.tsx [hunk 1]
```diff
@@ -183,167 +184,58 @@ export default function TabBuild() {
...
+    const buildTimer = setInterval(() => setBuildStatus({ type: "active", ...solver.computeStatus }), 100)
 
-      workers.push(worker)
-      cancelled.then(() => worker.terminate())
-      finalizedList.push(finalized)
-    }
-    for (let i = 0; i < 3; i++)
-      idleWorkers.push(...range(0, maxWorkers - 1))
+    const results = await solver.solve()
+    // END Solver
```
Comment (lantua) [inline apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabOptimize/index.tsx:229]: If it's this short, we probably don't need bracketing comments.

File: apps/frontend/src/app/Solver/SolverBase.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+          case "interim":
+            this.handleInterim(data as SourcedInterimResult)
+            break
+          default:
+            this.ipc(data as Result_t, i)
+        }
+        this.afterOnMessage()
+      }
+      this.workers.push(worker)
+      this.cancelled?.then(() => worker.terminate())
```
Comment (lantua) [inline apps/frontend/src/app/Solver/SolverBase.ts:126]: `cancelled` is no longer optional

**Review by lantua — APPROVED** [review]

**Review by nguyentvan7 — APPROVED** [review]

**Comment by tooflesswulf** [comment]
This fixes (https://github.com/frzyc/genshin-optimizer/issues/945) as well

**Review by frzyc — CHANGES_REQUESTED** [review]
Looks like you removed the windows cached version of esbuild, and commit in install-state.gz
clean up using 
`git checkout 7c1fd65cb9b060ebaf4b6361f02ba7d11c80f013 -- ".yarn/cache/@esbuild-win32-x64-npm-0.16.16-b88e3d7e7d-8.zip" `
and
`git checkout 7c1fd65cb9b060ebaf4b6361f02ba7d11c80f013 -- ".yarn/install-state.gz"`

**Comment by tooflesswulf** [comment]
OK did the checkout

**Review by frzyc — APPROVED** [review]

**Pull request merged**